### PR TITLE
Generate virtual Init methods

### PR DESCRIPTION
**Pull request by lgolding**

### CODE REVIEW

**Comment by lgolding** [comment]
@michaelcfanning is added to the review. #Closed

**Comment by lgolding** [comment]
@harleenkohli is added to the review. #Closed

File: src/Sarif/ToDotNet/ToDotNet.targets [hunk 1]
```diff
@@ -21,7 +21,7 @@
...
     <Message Importance="high" Text="Generating SARIF object model..." />
-    <Exec Command="&quot;$(JsonSchemaToDotNetPath)&quot; --schema-name Sarif --schema-file-path &quot;%(JsonSchemaFile.FullPath)&quot; --output-directory &quot;$(_ObjectModelOutputDirectory)&quot; --force-overwrite --namespace-name %(JsonSchemaFile.Namespace) --root-class-name %(JsonSchemaFile.RootClassName) --copyright-file-path &quot;%(JsonSchemaFile.CopyrightFilePath)&quot; --hints-file-path &quot;%(JsonSchemaFile.HintsFilePath)&quot; --generate-cloning-code" />
+    <Exec Command="&quot;$(JsonSchemaToDotNetPath)&quot; --schema-name Sarif --schema-file-path &quot;%(JsonSchemaFile.FullPath)&quot; --output-directory &quot;$(_ObjectModelOutputDirectory)&quot; --force-overwrite --namespace-name %(JsonSchemaFile.Namespace) --root-class-name %(JsonSchemaFile.RootClassName) --copyright-file-path &quot;%(JsonSchemaFile.CopyrightFilePath)&quot; --hints-file-path &quot;%(JsonSchemaFile.HintsFilePath)&quot; --generate-cloning-code --protected-init-methods" />
```
Comment (lgolding) [inline src/Sarif/ToDotNet/ToDotNet.targets]: >--protected-init-methods [](start = 466, length = 24)

The point of the exercise.

File: src/Sarif/Autogenerated/Artifact.cs [hunk 2]
```diff
@@ -212,7 +212,7 @@ private ISarifNode DeepCloneCore()
             return new Artifact(this);
         }
 
-        private void Init(Message description, ArtifactLocation location, int parentIndex, int offset, int length, ArtifactRoles roles, string mimeType, ArtifactContent contents, string encoding, string sourceLanguage, IDictionary<string, string> hashes, DateTime lastModifiedTimeUtc, IDictionary<string, SerializedPropertyInfo> properties)
+        protected virtual void Init(Message description, ArtifactLocation location, int parentIndex, int offset, int length, ArtifactRoles roles, string mimeType, ArtifactContent contents, string encoding, string sourceLanguage, IDictionary<string, string> hashes, DateTime lastModifiedTimeUtc, IDictionary<string, SerializedPropertyInfo> properties)
```
Comment (lgolding) [inline src/Sarif/Autogenerated/Artifact.cs]: >protected virtual [](start = 8, length = 17)

The result of the exercise.

File: src/ReleaseHistory.md [hunk 3]
```diff
@@ -185,23 +185,23 @@
 * Multitool improvements--added "rebaseUri" and "absoluteUri" tasks, which will either make the URIs in a SARIF log relative to some base URI, or take base URIs stored in SARIF and make the URIs absolute again.
 * Added a "processing pipeline" model to the SARIF SDK in order to allow easy chaining of operations on SARIF logs (like making all URIs relative/absolute).
 
-## **v2.0.0-csd.1** [Sdk](https://www.nuget.org/packages/Sarif.Sdk/2.0.0-csd.1) | [Driver](https://www.nuget.org/packages/Sarif.Driver/2.0.0-csd.1) | [Converters](https://www.nuget.org/packages/Sarif.Converters/2.0.0-csd.1)) | [Multitool](https://www.nuget.org/packages/Sarif.Multitool/2.0.0-csd.1))
```
Comment (lgolding) [inline src/ReleaseHistory.md]: >1)) [](start = 221, length = 3)

Scrolling through this I noticed a bunch of spurious trailing parentheses.

File: src/build.props [hunk 4]
```diff
@@ -11,7 +11,7 @@
     <Copyright Condition=" '$(Copyright)' == '' ">© Microsoft Corporation. All rights reserved.</Copyright>
     
     <!-- VersionPrefix denotes the current Semantic Version for the SDK. Must be updated before every nuget drop. -->
-    <VersionPrefix>2.1.3</VersionPrefix>
```
Comment (harleenkohli) [inline src/build.props]: >2.1.3 [](start = 19, length = 5)

we havent released 2.1.3 yet! so this change will go as part of 2.1.3, we should revert this change

File: src/ReleaseHistory.md [hunk 5]
```diff
@@ -375,17 +375,21 @@
...
 * Change schema uri to secure (https) instance.
 * BUGFIX: Fix tranformer bug where schema id would not be updated if no other transformation occurred.
 * BUGFIX: `threadFlowLocation.kind` value is getting lost during pre-release transformation. https://github.com/microsoft/sarif-sdk/issues/1502
-* BUGFIX: `location.logicalLocation` convenience setter mishandles null. https://github.com/microsoft/sarif-sdk/issues/1514
\ No newline at end of file
+* BUGFIX: `location.logicalLocation` convenience setter mishandles null. https://github.com/microsoft/sarif-sdk/issues/1514
+
+## **v2.1.4** [Sdk](https://www.nuget.org/packages/Sarif.Sdk/2.1.4) | [Driver](https://www.nuget.org/packages/Sarif.Driver/2.1.4) | [Converters](https://www.nuget.org/packages/Sarif.Converters/2.1.4) | [Multitool](https://www.nuget.org/packages/Sarif.Multitool/2.1.4)
+
+* API BREAKING: The Init methods in the Autogenerated SARIF object model classes are now virtual. This enables derived classes to add additional properties without having to copy the entire code of the Init method.
```
Comment (harleenkohli) [inline src/ReleaseHistory.md]: >* API BREAKING: The Init [](start = 0, length = 24)

this should go in section for 2.1.3

File: src/ReleaseHistory.md (hunk 3 above)
Comment (harleenkohli) [reply src/ReleaseHistory.md]: thanks! :)

---
In reply to: [290505867](https://github.com/microsoft/sarif-sdk/pull/1519#discussion_r290505867) [](ancestors = 290505867)

File: src/build.props (hunk 4 above)
Comment (lgolding) [reply src/build.props]: Done.

---
In reply to: [290513039](https://github.com/microsoft/sarif-sdk/pull/1519#discussion_r290513039) [](ancestors = 290513039)
